include/api.php

```diff
@@ -3276,72 +3276,706 @@ function api_oauth_access_token($type) {
+	 */
+	function api_fr_photo_create_update($type) {
+		if (api_user()===false)
 			throw new ForbiddenException();
-		} elseif (!x($_REQUEST, 'photo_id')) {
-			throw new BadRequestException("No photo id.");
+
+		// input params
+		$photo_id = (x($_REQUEST, 'photo_id') ? $_REQUEST['photo_id'] : null);
+		$desc = (x($_REQUEST, 'desc') ? $_REQUEST['desc'] : (array_key_exists('desc', $_REQUEST) ? "" : null)); // extra check necessary to distinguish between 'not provided' and 'empty string'
+		$album = (x($_REQUEST,'album') ? $_REQUEST['album'] : null);
+		$album_new = (x($_REQUEST,'album_new') ? $_REQUEST['album_new'] : null);
+		$allow_cid = (x($_REQUEST, 'allow_cid') ? $_REQUEST['allow_cid'] : (array_key_exists('allow_cid', $_REQUEST) ? " " : null));
+		$deny_cid = (x($_REQUEST, 'deny_cid') ? $_REQUEST['deny_cid'] : (array_key_exists('deny_cid', $_REQUEST) ? " " : null));
+		$allow_gid = (x($_REQUEST, 'allow_gid') ? $_REQUEST['allow_gid'] : (array_key_exists('allow_gid', $_REQUEST) ? " " : null));
+		$deny_gid = (x($_REQUEST, 'deny_gid') ? $_REQUEST['deny_gid'] : (array_key_exists('deny_gid', $_REQUEST) ? " " : null));
+		$visibility = (x($_REQUEST, 'visibility') ? (($_REQUEST['visibility'] == "true" || $_REQUEST['visibility'] == 1) ? true : false) : false);
+
+		// do several checks on input parameters
+		// we do not allow calls without album string
+		if ($album == null)
+			throw new BadRequestException("no albumname specified");
+
+		// if photo_id == null --> we are uploading a new photo
+		if ($photo_id == null) {
+			$mode = "create";
+
+			// error if no media posted in create-mode
+			if(!x($_FILES,'media')) {
+				// Output error
+				throw new BadRequestException("no media data submitted");
+			}
+
+			// album_new will be ignored in create-mode
+			$album_new = "";
+		} else {
+			$mode = "update";
+
+			// check if photo is existing in database
+			$r = q("SELECT * FROM `photo` WHERE `uid` = %d AND `resource-id` = '%s' AND `album` = '%s'",
+					intval(api_user()),
+					dbesc($photo_id),
+					dbesc($album));
+			if (!dbm::is_result($r))
+				throw new BadRequestException("photo not available");
+		}
+
+		// checks on acl strings provided by clients
+		$acl_input_error = false;
+		$acl_input_error |= CheckACLInput($allow_cid);
+		$acl_input_error |= CheckACLInput($deny_cid);
+		$acl_input_error |= CheckACLInput($allow_gid);
+		$acl_input_error |= CheckACLInput($deny_gid);
+		if ($acl_input_error)
+			throw new BadRequestException("acl data invalid");
+
+		// now let's upload the new media in create-mode
+		if ($mode == "create") {
+			$media = $_FILES['media'];
+			$data = SaveMediaToDatabase("photo", $media, $type, $album, trim($allow_cid), trim($deny_cid), trim($allow_gid), trim($deny_gid), $desc, $visibility);
+
+			// return success of updating or error message
+			if (!is_null($data)) {
+				return api_format_data("photo_create", $type, $data);
+			}
+			else {
+				throw new BadRequestException("unknown error");
```

Comment:
changed to "unknown error - uploading photo failed, see Friendica log for more information"